### ISSUE TITLE
Request: File limits/pagination

### ISSUE THREAD

**Issue by vrdriver** [Jako/FileDownloadR#8]
Hi Thomas,

I have about 1000 files I'm showing for downloads, but it takes a little while to load.

I was wondering if you could add a feature where it could only show the last 100 files, and therefore we use pagination to cycle through the rest of the files?

Cheers,
Steve

**Comment by Jako** [comment]
It is not that easy. You have to split the $sort['file'] array by limit/offset somewhere inside of https://github.com/Jako/FileDownloadR/blob/master/core/components/filedownloadr/src/FileDownloadR.php#L1214-L1221

Then the snippet has to be prepared to be usable by getPage/pdoPage (limit/offset properties + total placeholder).

Also the pagination should only be possible when the list shows only one folder.

**Comment by Jako** [comment]
Can you check the 3.2.0 RC1 from https://github.com/Jako/FileDownloadR/tree/3.2.0/_packages, please?

**Comment by vrdriver** [comment]
Thanks. I'll hopefully try in the next few days.

**Comment by vrdriver** [comment]
Hi Jako, apologies for the delay. I've been super busy at work, and am only just trying this now. 

I'm trying it on a local dev server before I deploy it to the real server.  
So far, I've been able to migrate my live site to a local dev server and the older version (3.1.4-pl) is working fine, but this new version isn't playing the game correctly for me at the moment.   
In my process, I removed the previous version to allow me to put on version 3.2.0-rc2 manually from downloading the transport package and choosing to "Upload Package" and letting modx to the install.

However, after doing this, simply put, it's not displaying any files in the list. 

I know it's seeking and reading the files, because when I give it the correct path, the page takes longer to load, but I just get an empty document. 

This is what I'm calling for testing.

`[[!FileDownload?`  
``getDir=`../distribution/audio_files` ``   
``&browseDirectories = `0` ``   
``&showEmptyDirectory = `0` ``   
``&extShown=`doc,mp3,docx` ``   
``&sortOrder' = `asc` ``   
`]]`

And in Live production I'm calling the snippet programmatically like so.  

`echo $modx->runSnippet('FileDownload', [`  
`'getDir' => $_path,`  
`'browseDirectories' => '0',`  
`'showEmptyDirectory' => '1',`  
`'extShown' => 'doc,mp3,docx',`  
`'sortOrder' => $_sortOrder`  
`]);`

However, both are giving the same results.  

![](https://github.com/user-attachments/assets/fb47d6e7-2c52-4a19-a404-5393234348de)

Unfortunately there is nothing showing in the logs, of apache or php or modx for your snippet.

I'm not sure what else needs to happen.  
  
MODX Version: MODX Revolution 3.0.4-pl (traditional)
And I've tried multiple versions of php on my local dev MAMP set up - all with the same result.

**Comment by Jako** [comment]
Can you try the RC3, please.

**Comment by vrdriver** [comment]
I just tried, and still a similar result.


However, it showed the path name this time.

![Screenshot 2025-01-11 at 12 02 37](https://github.com/user-attachments/assets/861c4019-ac71-4f9c-b15a-4cb72298f123)



![Screenshot 2025-01-11 at 12 03 03](https://github.com/user-attachments/assets/9f7c5b1d-0a38-428d-b473-1d3b64c9a110)

This time I have saved the output from the install as well, which may help?


```

/*
* MODX Console Output
*
* @date 2025-01-11 11:55:32 */
Attempting to install package with signature: filedownloadr-3.2.0-rc3 Package found...now preparing to install.
Grabbing package workspace...
Workspace environment initiated, now installing package...
Installing/updating dependencies, this may take some time...
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/installed.php): failed to open stream: No such file or directory
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/InstalledVersions.php): failed to open stream: No such file or directory 
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/autoload_namespaces.php): failed to open stream: No such file or directory
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/autoload_psr4.php): failed to open stream: No such file or directory
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/autoload_classmap.php): failed to open stream: No such file or directory 
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/autoload_static.php): failed to open stream: No such file or directory 
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/platform_check.php): failed to open stream: No such file or directory
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/autoload.php): failed to open stream: No such file or directory 
PHP warning: file_get_contents(/Users/stephenmonro/Sites/test/core/components/filedownloadr/vendor/composer/autoload_real.php): failed to open stream: No such file or directory Installing dependencies from lock file Verifying lock file contents can be installed on current platform.
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update `.
Package operations: 2 installs, 0 updates, 0 removals - 
Downloading ip2location/ipinfodb-php (1.0.0) - 
Downloading ralouphie/mimey (2.1.0) - 
Installing ip2location/ipinfodb-php (1.0.0): 
Extracting archive - Installing ralouphie/mimey (2.1.0): 
Extracting archive Generating optimized autoload files No security vulnerability advisories found.
Altering table: fdCount - Updating columns
-- altered column: id
-- altered column: ctx
-- altered column: filename -- altered column: count -- altered column: hash - Updating indexes
Altering table: fdDownloads - Updating columns
-- altered column: id
-- altered column: path_id -- altered column: ip
-- altered column: referer -- altered column: country -- altered column: region -- altered column: city -- altered column: zip
-- altered column: geolocation -- altered column: user -- altered column: timestamp - Updating indexes
-- removed index: path_id -- removed index: user
-- added index: path_id -- added index: user
Altering table: fdPaths - Updating columns
-- altered column: id
-- altered column: ctx
-- altered column: media_source_id -- altered column: filename -- altered column: extended -- altered column: hash - Updating indexes
System event "OnFileDownloadLoad" was created.
System event "OnFileDownloadBeforeDirOpen" was created.
System event "OnFileDownloadAfterDirOpen" was created.
System event "OnFileDownloadBeforeFileDownload" was created.
System event "OnFileDownloadAfterFileDownload" was created.
System event "OnFileDownloadBeforeFileUpload" was created.
System event "OnFileDownloadAfterFileUpload" was created.
System event "OnFileDownloadBeforeFileRemove" was created.
System event "OnFileDownloadAfterFileRemove" was created.
Successfully installed package filedownloadr-3.2.0-rc3 
/* EOF */
```

**Comment by Jako** [comment]
Can you post the snippet call please?

**Comment by vrdriver** [comment]
Absolutely. Here's the snippet, and related screen shots of what I have as well.

    
[[!FileDownload?
getDir=`../audio/files/`
&browseDirectories = `0`
&showEmptyDirectory = `0`
&extShown=`doc,mp3,docx` 
&sortOrder' = `asc`
]]

![Screenshot 2025-01-11 at 23 13 17](https://github.com/user-attachments/assets/8ceb57d8-6bef-4086-9ea1-88101211fd53)

Here is the output:
![Screenshot 2025-01-11 at 23 13 35](https://github.com/user-attachments/assets/957a1949-dc47-4d16-9da2-34171637e313) 

Here's the list of the files we are trying to reach, which are outside the main root path of the modx site.
Main modx root for this is:
/Users/stephenmonro/sites/test/

![Screenshot 2025-01-11 at 23 19 07](https://github.com/user-attachments/assets/bfe5bcaf-3dad-4d12-afea-85b47fdf3746)

The snippet code, as per the transport package as well:
![Screenshot 2025-01-11 at 23 22 10](https://github.com/user-attachments/assets/74467c34-cbfc-4842-951c-3a4c40ec0afa)

Here's the package as installed:
![Screenshot 2025-01-11 at 23 14 04](https://github.com/user-attachments/assets/519efb5c-74f9-46d1-81f2-cc2541786251)

It's currently on PHP Version 7.4.33, but I can push that a little if required.
I even tried removing the .htaccess file and referencing the file by index.php?id=9 and still the same result.
I could safely update to the latest version of MODX if it helps?

**Comment by Jako** [comment]
This should be solved with version 3.2.0.
